Add support for generics

SoMuchForSubtlety opened the pull request:
This PR adds support for generics. 

Because there is no generic list type in the standard library yet, I added a minimal version of `container/list` with added generics support (I kept the licence header, please let me know if it should be removed).

It would also be possible to keep using `container/list` as is, and just update the public API (take care of interface casts internally), but most of the performance gain comes from removing interface conversions.

Benchmark results against master
```
name         old time/op    new time/op    delta
2Q_Rand-16     1.22µs ±10%    0.52µs ± 8%   -57.69%  (p=0.000 n=20+19)
2Q_Freq-16     1.03µs ±10%    0.44µs ±12%   -57.42%  (p=0.000 n=18+20)
ARC_Rand-16    1.51µs ± 9%    0.70µs ±15%   -53.63%  (p=0.000 n=20+20)
ARC_Freq-16    1.22µs ±12%    0.54µs ± 7%   -56.20%  (p=0.000 n=20+20)
LRU_Rand-16     401ns ± 9%     215ns ± 6%   -46.46%  (p=0.000 n=20+20)
LRU_Freq-16     376ns ±13%     194ns ± 7%   -48.51%  (p=0.000 n=19+20)

name         old alloc/op   new alloc/op   delta
2Q_Rand-16       136B ± 0%       67B ± 0%   -50.74%  (p=0.000 n=20+18)
2Q_Freq-16       124B ± 0%       59B ± 0%   -52.42%  (p=0.000 n=16+20)
ARC_Rand-16      165B ± 0%       84B ± 1%   -49.03%  (p=0.000 n=20+20)
ARC_Freq-16      139B ± 3%       68B ± 5%   -51.01%  (p=0.000 n=20+20)
LRU_Rand-16     76.0B ± 0%     36.0B ± 0%   -52.63%  (p=0.000 n=20+20)
LRU_Freq-16     71.0B ± 0%     33.0B ± 0%   -53.52%  (p=0.000 n=19+20)

name         old allocs/op  new allocs/op  delta
2Q_Rand-16       5.00 ± 0%      1.00 ± 0%   -80.00%  (p=0.000 n=20+20)
2Q_Freq-16       5.00 ± 0%      1.00 ± 0%   -80.00%  (p=0.000 n=20+20)
ARC_Rand-16      6.00 ± 0%      1.00 ± 0%   -83.33%  (p=0.000 n=20+20)
ARC_Freq-16      5.00 ± 0%      1.00 ± 0%   -80.00%  (p=0.000 n=20+20)
LRU_Rand-16      3.00 ± 0%      0.00       -100.00%  (p=0.000 n=20+20)
LRU_Freq-16      3.00 ± 0%      0.00       -100.00%  (p=0.000 n=20+20)
```

jefferai commented on go.mod:
```diff
@@ -1,3 +1,3 @@
 module github.com/hashicorp/golang-lru
 
-go 1.12
+go 1.19
```
Two things here:

1) The module name needs to be updated to `github.com/hashicorp/golang-lru/v2`
2) Is Go 1.19 specifically needed or does Go 1.18 suffice? I think it's fine in `/v2` to require 1.19 if it's really needed but wanted to check.

SoMuchForSubtlety replied:
I changed it to 1.18

jefferai commented on 2q.go:
```diff
@@ -103,11 +103,11 @@ func (c *TwoQueueCache) Get(key interface{}) (value interface{}, ok bool) {
 
 	// No hit
-	return nil, false
+	return
```
Elsewhere in this function it explicitly returns parameters instead of using the named parameters. While this function will work correctly so far as I know, I think it reads more confusing that it needs to. Since you're updating stuff, can you remove the named return parameters (here and in other Get methods) and keep the `return nil, false`? It's more clear/explicit.

jefferai replied:
You could also keep the named params for clarity in documentation but just be explicit in all return clauses.

SoMuchForSubtlety replied:
We can't keep `return nil, false`, because `nil` can't be assigned to all possible types of `V` (privimite types, non-pointer structs, etc.). This is also the reason for the other changes to naked returns.

https://go.dev/play/p/tf-1bsx02Q1

I can change the other return statements or add comments if you want.

jefferai replied:
Interesting. Given that `any` is the same as `interface{}`, which itself is a reference to some other type but can always be nil, I wonder why generics behave this way.

For instance it is valid to have `var p interface{}` which is equivalent to `nil`. It is also valid to have `var b *Foobar; var p interface{}; p = b` which would be a nil pointer inside a now not-nil interface.

I guess I assumed that `any` in this context for generics meant the same as `any` in non-generic uses.

So it seems like there are two options: one is accept that some types may return empty instead of nil, document this as a breaking change in v2, and tell people if they want the old semantics to always use correct types; the other is to have generics for K only and just use `any` as the value type in all cases. Probably better to do the former though.

SoMuchForSubtlety replied:
I don't think this is a breaking change. Returning empty values is fine because `ok` will be false, indicating that there was no hit and the return value isn't valid. The built-in go map has the exact same behaviour[^1].

We could modify the function signature to make it return a pointer[^2], but I don't really like that approach.

[^1]: https://go.dev/play/p/MohR-gp_zgs
[^2]: https://go.dev/play/p/zLL4NU2tfyX

jefferai replied:
Ah yes, forgot about `ok` there

jefferai commented on arc.go:
```diff
@@ -60,7 +60,7 @@ func NewARC(size int) (*ARCCache, error) {
 
 // Get looks up a key's value from the cache.
-func (c *ARCCache) Get(key interface{}) (value interface{}, ok bool) {
+func (c *ARCCache[K, V]) Get(key K) (value V, ok bool) {
```
Same comment in this get function

jefferai commented on arc.go:
```diff
@@ -177,30 +177,32 @@ func (c *ARCCache) Add(key, value interface{}) {
 		k, _, ok := c.t1.RemoveOldest()
 		if ok {
-			c.b1.Add(k, nil)
+			var empty V
```
Can you do me a favor and just explain this change? I haven't worked with generics a lot, but since `V` is of type `any`, isn't it perfectly valid to continue using `nil` here?

This will still compare against `nil` (tested it for my own benefit, https://go.dev/play/p/PH5XDYNs-Qu). So I don't think it's wrong, but curious about it.

SoMuchForSubtlety replied:
We can't keep `return nil, false`, because `nil` can't be assigned to all possible types of `V` (privimite types, non-pointer structs, etc.). This is also the reason for the other changes to naked returns.

https://go.dev/play/p/tf-1bsx02Q1

I can change the other return statements or add comments if you want.

jefferai commented on lru.go:
```diff
@@ -136,8 +140,9 @@ func (c *Cache) ContainsOrAdd(key, value interface{}) (ok, evicted bool) {
 // Returns whether found and whether an eviction occurred.
-func (c *Cache) PeekOrAdd(key, value interface{}) (previous interface{}, ok, evicted bool) {
-	var k, v interface{}
+func (c *Cache[K, V]) PeekOrAdd(key K, value V) (previous interface{}, ok, evicted bool) {
```
```suggestion
func (c *Cache[K, V]) PeekOrAdd(key K, value V) (previous any, ok, evicted bool) {
```

jefferai commented on simplelru/list.go:
```diff
@@ -0,0 +1,128 @@
+// Copyright 2009 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
```
Probably this license should be added; can you copy over the LICENSE file and name it LICENSE_list and update the comment here to point to it? And at the top of the LICENSE_list file add a comment indicating it applies only to list.go?

jefferai commented on simplelru/lru.go:
```diff
@@ -70,37 +63,34 @@ func (c *LRU) Add(key, value interface{}) (evicted bool) {
+		return ent.value, true
 	}
-	return nil, ok
+	return
```
Same comments about return values in this file. We don't seem to use the named params so I'd prefer we just get rid of them.

jefferai replied:
Or keep them but be explicit in all returns.

kasegao commented:
Thank you for your great PR.

In order to keep backwward compability, however, wouldn't it be better to keep the existing implementation as suggested by AskAlexSharov's  comment?

https://github.com/hashicorp/golang-lru/pull/101#issuecomment-1269275609

jefferai commented:
@kasegao If this is marked as v2 of the lib we don't need backwards compat.

kasegao commented:
@jefferai I misunderstood. Thank you!

SoMuchForSubtlety commented:
golangci-lint fails because of a few false positives. I could disable the offending linters (golint and gosec).
https://github.com/SoMuchForSubtlety/golang-lru/actions/runs/3347498929

There are warnings about other enabled linters being deprecated
```
WARN [runner] The linter 'structcheck' is deprecated (since v1.49.0) due to: The owner seems to have abandoned the linter.  Replaced by unused. 
WARN [runner] The linter 'deadcode' is deprecated (since v1.49.0) due to: The owner seems to have abandoned the linter.  Replaced by unused. 
WARN [runner] The linter 'scopelint' is deprecated (since v1.39.0) due to: The repository of the linter has been deprecated by the owner.  Replaced by exportloopref. 
WARN [runner] The linter 'varcheck' is deprecated (since v1.49.0) due to: The owner seems to have abandoned the linter.  Replaced by unused. 
WARN [linters_context] structcheck is disabled because of generics. You can track the evolution of the generics support by following the https://github.com/golangci/golangci-lint/issues/2649. 
```

jefferai commented:
I just bumped the workflows version in CI in `main`, you may want to try updating from there and seeing if that helps.

SoMuchForSubtlety commented:
> I just bumped the workflows version in CI in `main`, you may want to try updating from there and seeing if that helps.

Rebasing on master gets rid of the `math/rand` warnings, but I still get the errors from `golint`. It's deprecated[^1] and does not have support for generics, leading to it getting confused by generic method signatures.

[^1]: https://github.com/golang/go/issues/38968

jefferai commented:
I removed some deprecated linters from the config -- didn't even realize we had a config and weren't using defaults. Hopefully that fixes it.

SoMuchForSubtlety commented:
> I removed some deprecated linters from the config -- didn't even realize we had a config and weren't using defaults. Hopefully that fixes it.

Thanks, I rebased on master and at least locally golangci-lint doesn't complain any more.

Is there anything else open I need to change?

jefferai commented:
I will do another pass -- as for stuff outstanding I still have a slight preference not to do bare `return` when nil/default values are being returned from Get functions considering we're not actually populating the named return parameters anywhere. Keeping them named is nice from a doc perspective, but we could explicitly return the false/nil values. But it's more a style thing than anything else. 🤷

SoMuchForSubtlety commented:
> Keeping them named is nice from a doc perspective, but we could explicitly return the false/nil values.

Do you mean
```go
// Peek returns the key value (or undefined if not found) without updating
// the "recently used"-ness of the key.
func (c *LRU[K, V]) Peek(key K) (value V, ok bool) {
	var ent *entry[K, V]
	if ent, ok = c.items[key]; ok {
		return ent.value, true
	}
	return value, false
}
```
or 
```go
// Peek returns the key value (or undefined if not found) without updating
// the "recently used"-ness of the key.
func (c *LRU[K, V]) Peek(key K) (value V, ok bool) {
	var ent *entry[K, V]
	if ent, ok = c.items[key]; ok {
		return ent.value, true
	}
	var empty V
	return empty, false
}
```

They are the same, but the seconds one makes it clearer that we are returning an empty value (at the cost of an added line).

jefferai commented:
Hm. I keep forgetting that V is not necessarily `nil` anymore :-) 

Leave it be.

jefferai commented:
Looks great! I have a question for anyone following this:

I'm tempted to cut a final `v0` tag, set this as `v2`, and have `v1` point to a branch that has a README that basically says, use v0 if you need backwards compat or use v2. This would break users that auto-upgrade until they select a specific v0 (or v2) tag, but also pushes people towards using v2 given that it's faster and more capable.

Thoughts?

bep commented:
>Thoughts?

I agree (I do, though, wish Go Modules had a way to handle this cleaner).

jefferai commented:
@SoMuchForSubtlety Any thoughts/preferences on my above question?

SoMuchForSubtlety commented:
Sounds reasonable, I have no objections :)

jefferai commented:
Amazing work! Thanks!

Kubuxu commented on arc.go:
```diff
@@ -14,41 +14,41 @@ import (
-	t1 simplelru.LRUCache // T1 is the LRU for recently accessed items
-	b1 simplelru.LRUCache // B1 is the LRU for evictions from t1
+	t1 simplelru.LRUCache[K, V] // T1 is the LRU for recently accessed items
+	b1 simplelru.LRUCache[K, V] // B1 is the LRU for evictions from t1
```
As the eviction tracking LRU caches don't use the value, whouldn't it be better to use `LRUCache[K, struct{}]` here? I might be missing something, but I can open a PR if I don't.
